File: log-generator/log_generator.py
# to run : python log_generator.py 

import requests
import random
import uuid
from datetime import datetime, timedelta, timezone

BASE_URL = "http://localhost:8080/api/addLogs"

SERVICES = ["auth-service", "payment-service", "order-service", "inventory-service", "notification-service"]
LEVELS = ["DEBUG", "INFO", "WARNING", "ERROR", "FATAL"]
MESSAGES = [
    "User login successful",
    "Payment processed",
    "Database connection timeout",
    "Order created",
    "Cache miss for key",
    "Null pointer exception in handler",
    "Retry attempt failed",
    "Request completed successfully",
]

def generate_log(base_time):
    ts = base_time - timedelta(seconds=random.randint(0, 60 * 60 * 24 * 7))  # spread over last 7 days
    return {
        "timeStamp": ts.isoformat().replace("+00:00", "Z"),
        "serviceName": random.choice(SERVICES),
        "logLevel": random.choices(LEVELS, weights=[30, 40, 15, 10, 5])[0],  # INFO/DEBUG more common
        "message": random.choice(MESSAGES),
        "source": random.choice(["Postman", "LoadGenerator", "InternalService"]),
        "metadata": f'{{"requestId":"{uuid.uuid4()}"}}'
    }

def generate_batch(size):
    now = datetime.now(timezone.utc)
    return [generate_log(now) for _ in range(size)]

def send_bulk(total=5000, batch_size=200):
    sent = 0
    while sent < total:
        current_batch = min(batch_size, total - sent)
        payload = generate_batch(current_batch)
        resp = requests.post(BASE_URL, json=payload)
        if resp.status_code == 200:
            sent += current_batch
            print(f"Sent {sent}/{total}")
        else:
            print(f"Failed batch: {resp.status_code} - {resp.text}")
            break

if __name__ == "__main__":
    send_bulk(total=10000, batch_size=200)

File: frontend/src/lib/logGenerator.js
const SERVICES = [
  "auth-service",
  "payment-service",
  "order-service",
  "inventory-service",
  "notification-service",
];
const LEVELS = ["DEBUG", "INFO", "WARNING", "ERROR", "FATAL"];
const LEVEL_WEIGHTS = [30, 40, 15, 10, 5]; // INFO/DEBUG more common than errors, like real traffic
const MESSAGES = [
  "User login successful",
  "Payment processed",
  "Database connection timeout",
  "Order created",
  "Cache miss for key",
  "Null pointer exception in handler",
  "Retry attempt failed",
  "Request completed successfully",
];

function weightedRandom(items, weights) {
  const total = weights.reduce((a, b) => a + b, 0);
  let r = Math.random() * total;
  for (let i = 0; i < items.length; i++) {
    r -= weights[i];
    if (r <= 0) return items[i];
  }
  return items[items.length - 1];
}

function randomTimestamp(daysBack) {
  const offsetMs = Math.random() * daysBack * 24 * 3600 * 1000;
  return new Date(Date.now() - offsetMs);
}

function buildLog(timestamp, levelOverride) {
  return {
    timeStamp: timestamp.toISOString(),
    serviceName: SERVICES[Math.floor(Math.random() * SERVICES.length)],
    logLevel: levelOverride || weightedRandom(LEVELS, LEVEL_WEIGHTS),
    message: MESSAGES[Math.floor(Math.random() * MESSAGES.length)],
    source: "AppGenerator",
    metadata: JSON.stringify({ requestId: crypto.randomUUID() }),
  };
}

/**
 * Generates `count` logs spread over the last `daysBack` days, plus a
 * few injected "incident" windows (short bursts of ERROR/FATAL) so the
 * trend and error-rate charts have something visually interesting.
 */
export function generateLogs(count, { daysBack = 7, incidentCount = 3 } = {}) {
  const burstSize = Math.max(5, Math.round(count * 0.01));
  const reserved = burstSize * incidentCount;
  const baselineCount = Math.max(0, count - reserved);

  const logs = [];
  for (let i = 0; i < baselineCount; i++) {
    logs.push(buildLog(randomTimestamp(daysBack)));
  }

  for (let i = 0; i < incidentCount; i++) {
    const windowStart =
      Date.now() - Math.random() * daysBack * 24 * 3600 * 1000;
    const windowMs = 45 * 60 * 1000;
    for (let j = 0; j < burstSize; j++) {
      const ts = new Date(windowStart + Math.random() * windowMs);
      logs.push(buildLog(ts, Math.random() < 0.6 ? "ERROR" : "FATAL"));
    }
  }
  return logs;
}

export function chunk(array, size) {
  const chunks = [];
  for (let i = 0; i < array.length; i += size) {
    chunks.push(array.slice(i, i + size));
  }
  return chunks;
}
